File: 2015/dia02/main.py
from functools import reduce
from operator import mul

input = open("input.txt", "r").read().strip()
print(input)


def parte_1():
    total_square_feet = 0
    for line in input.split():
        dims = [int(d) for d in line.split("x")]
        faces_square_feet = [dims[0] * dims[1], dims[0] * dims[2], dims[1] * dims[2]]
        total_square_feet += 2 * sum(faces_square_feet) + min(faces_square_feet)
    print(total_square_feet)


def parte_2():
    total_feet = 0
    for line in input.split():
        dims = [int(d) for d in line.split("x")]
        sorted_dims = sorted(dims)
        feet_1 = 2 * (sorted_dims[0] + sorted_dims[1])
        feet_2 = reduce(mul, dims, 1)
        total_feet += feet_1 + feet_2
    print(total_feet)


if __name__ == "__main__":
    parte_1()
    parte_2()

File: 2015/dia02/cpp/main.cpp
#include <algorithm>
#include <fstream>
#include <numeric>
#include <print>
#include <ranges>
#include <vector>

namespace rg = std::ranges;
namespace vw = std::views;

using std::string;

std::string to_string(std::ranges::view auto a) {
    // return std::string{a.begin(), a.end()};
    return std::string{std::from_range, a};
}

void parte_1(std::vector<std::string>& lines) {
    int total = 0;
    for (std::string_view line : lines) {
        auto splitted_line =
            line | vw::split('x') |
            vw::transform([](auto a) { return std::stoi(to_string(a)); }) |
            rg::to<std::vector<int>>();
        int a = splitted_line[0];
        int b = splitted_line[1];
        int c = splitted_line[2];
        // std::print("{} {} {}\n", a, b, c);
        std::vector faces_square_feet = {a * b, a * c, b * c};
        int total_square_feet = 2 * std::reduce(faces_square_feet.begin(),
                                                faces_square_feet.end(), 0) +
                                *rg::min_element(faces_square_feet);
        total += total_square_feet;
    }
    std::print("Total parte 1: {}\n", total);
};

void parte_2(std::vector<std::string>& lines) {
    int total = 0;
    for (std::string_view line : lines) {
        auto splitted_line =
            line | vw::split('x') | vw::transform([](auto a) {
                return std::stoi(std::string{std::from_range, a});
            }) |
            rg::to<std::vector<int>>();
        rg::partial_sort(splitted_line.begin(), splitted_line.begin() + 2,
                         splitted_line.end());
        int a = splitted_line[0];
        int b = splitted_line[1];
        // int c = splitted_line[2];
        // std::print("{} {} {}\n", a, b, c);
        int feet_1 = 2 * (a + b);
        int feet_2 = std::reduce(splitted_line.begin(), splitted_line.end(), 1,
                                 std::multiplies<>{});
        total += feet_1 + feet_2;
    }
    std::print("Total parte 2: {}\n", total);
};

int main(int argc, char** argv) {
    std::print("\n");

    std::ifstream input;
    input.open(argv[1]);

    std::string line;
    std::vector<std::string> lines;
    while (std::getline(input, line)) {
        lines.push_back(line);
    }
    input.close();

    parte_1(lines);
    parte_2(lines);

    return 0;
}
